text/0000-deprecate-string-based-actions.md

```diff
@@ -0,0 +1,89 @@
+- Start Date: 2020-05-23
+- Relevant Team(s): Ember.js
+- RFC PR: (after opening the RFC PR, update this with a link to it and update the file name)
+- Tracking: (leave this empty)
+
+# Deprecate String Based Actions
+
+## Motivation
+
+> In Ember 1.13 we introduced [closure actions](https://github.com/emberjs/rfcs/blob/00ac2685c86f27d41547012903f485a4ef338d27/active/0000-improved-actions.md) have been recommended over string based actions. This officially deprecates string based actions in favor of closure actions,
+which provide many benefits such as
+
+1. simpler passing through components
+2. less need for an actions hash
+3. ability to return a value
+4. typing in TypeScript
+5. compatible with Glimmer Components
+
+## Detailed Design
+
+We will deprecate the following:
+
+1. The `send` api.
+2. The use of a string parameter to the `action` helper.
+
+## Transition Path
+
+Users will transition to closure actions.
+
+From:
+
+```js
+export default Component.extend({
+    someMethod() {
+        this.send('actionName');
+    }
+
+    actions: {
+        actionName() {
+            // do something
+        }
+    }
+});
+```
+
+To:
+
+```js
+export default Component.extend({
+    someMethod() {
+        this.actionName();
+    }
+
+    actionName() {
```

Comment:
This should use `@action`

Reply:
`@action` will not work with `.extend(`; i.e. it would require a native class.

Reply:
I'm using `@action` here to show what I mean, you can easily use that decorator in a classic Ember.Object.extend style class (just like computed):

```js
export default Component.extend({
  actionName: action(function() {

  }),
})
```

Reply:
I'm doubtful we should be recommending this.

Reply:
It was recommended in the 3.11 release blog post https://blog.emberjs.com/2019/07/15/ember-3-11-released.html, we used this syntax pretty extensively while migrating!

Reply:
If there is a gap in the guides (though I’m not sure there is) then we should fix it. That is something this RFC should work to address. 

That fact does **not** change the direction we should advise folks to migrate to. Migrating as proposed in this RFC at the moment would be a non-trivial waste of time, and ultimately (IMO) not worth moving forward RFC wise.

Reply:
Maybe this RFC / deprecation guide can say something like:

> While the action helper and modifiers are still supported with non-string arguments, the refactoring recommendation is to switch to the `on` modifier and `@action` decorator directly. It is likely that the `action` helpers and modifiers will be deprecated in future RFCs.

and give an example like this:

> For example, take this button that sends the `foo` action with the default click event:

> ```hbs
> <button {{action "foo"}}></button>
> ```

> Although this _could_ be refactored to:

> ```hbs
> <button {{action this.foo}}></button>
> ```

> where `foo()` is implemented in the backing context in the actions hash:

> ```js
> Component.extend({
>   actions: {
>     foo() {}
>   }
> })
> ```

> it is recommended that the code is refactored instead to:

> ```hbs
> <button {{on "click" this.foo}}></button>
> ```

> ```js
> Component.extend({
>   foo: action(function() {})
> })
> ```

This could also be written as the with the `diff` syntax to make it more clear what changed. It doesn't quite capture the case of refactoring away from an action that bubbles, but you could link to a page in the guides that covers it.

Reply:
Ok. I'm putting this in. Where in the guides can we discuss this?

Reply:
Not 100% sure, maybe a "classic interop" thing in the advanced topics? 

@emberjs/learning-team-managers - Thoughts?